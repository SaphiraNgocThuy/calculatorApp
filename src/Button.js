import React, { memo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { ButtonView, StyledButton, ButtonText } from './styles';
import { useDispatch } from 'react-redux';
import { addNumber, addOperator, clear, getResult, addDot } from './actions';

const Button = ({ title, isLeftSide }) => {
  const dispatch = useDispatch();

  const onPress = useCallback(() => {
    if (/\d/.test(title)) {
      addNumber(dispatch, title);
    } else if (/[\+\-\*\/]/.test(title)) {
      addOperator(dispatch, title);
    } else if (/\=/.test(title)) {
      getResult(dispatch);
    } else if (/C/.test(title)) {
      clear(dispatch);
    } else {
      addDot(dispatch);
    }
  }, [title]);

  return (
    <StyledButton {...{ onPress, isLeftSide }}>
      <ButtonText>{title}</ButtonText>
    </StyledButton>
  );
};

Button.propTypes = {
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isLeftSide: PropTypes.bool,
};

export default memo(Button);
