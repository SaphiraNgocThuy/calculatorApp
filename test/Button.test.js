import React from 'react';
import * as redux from 'react-redux';
import renderer from 'react-test-renderer';
import Button from '../src/Button';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
describe('actions', () => {
  it('renders correctly', () => {
    const useDispatch = jest.spyOn(redux, 'useDispatch');
    useDispatch.mockReturnValue(jest.fn());

    const tree = renderer.create(<Button title="1" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('call add number if press number button ', () => {
    Enzyme.configure({ adapter: new Adapter() });
    const useDispatch = jest.spyOn(redux, 'useDispatch');
    useDispatch.mockReturnValue(jest.fn());
    const addNumber = jest.fn();
    const dispatch = jest.fn();
    const title = '1';
    const button = shallow(<Button title={title} />);
    button.props().onPress();

    expect(addNumber).toHaveBeenCalledWith(dispatch, title);
  });
});
