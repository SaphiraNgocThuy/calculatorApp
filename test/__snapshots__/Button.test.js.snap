// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "lightgray",
      "borderRadius": 5,
      "flexBasis": "29%",
      "flexGrow": 1,
      "flexShrink": 1,
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 5,
      "paddingLeft": 5,
      "paddingRight": 5,
      "paddingTop": 5,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    1
  </Text>
</View>
`;
